### PR TITLE
Fix #17741 (return type of sets built form generators)

### LINKED ISSUE #17741: The type of `Set(x for x in 1:10)` is inferred as `Set{Any}`.
```julia
julia> Set(x for x in 1:10)
Set(Any[7,4,9,10,2,3,5,8,6,1])

julia> Set(x for x in 1:10) |> typeof
Set{Any}

```

Apparently, the expected type is `Set{Int}`.

---

```
julia> versioninfo()
Julia Version 0.5.0-rc0+86
Commit d68448c* (2016-08-01 03:45 UTC)
Platform Info:
  System: Darwin (x86_64-apple-darwin14.5.0)
  CPU: Intel(R) Core(TM) i5-4288U CPU @ 2.60GHz
  WORD_SIZE: 64
  BLAS: libopenblas (USE64BITINT DYNAMIC_ARCH NO_AFFINITY Haswell)
  LAPACK: libopenblas64_
  LIBM: libopenlibm
  LLVM: libLLVM-3.7.1 (ORCJIT, haswell)

```

### PR DESCRIPTION
**Pull request by p-zubieta**

### CODE REVIEW

File: test/sets.jl [hunk 1]
```diff
@@ -11,6 +11,12 @@ data_out = collect(s)
 @test is(typeof(data_out), Array{Any,1})
 @test all(map(d->in(d,data_out), data_in))
 @test length(data_out) == length(data_in)
+let f17741 = x -> x < 0 ? false : 1
+    @test isa(Set(x for x = 1:3), Set{Int})
```
Comment (KristofferC) [inline test/sets.jl]: Does `@inferred` work?

Comment (pabloferz) [inline test/sets.jl]: Yeah, it does. Should I use it here?

Comment (KristofferC) [inline test/sets.jl]: Maybe nice to test type stability as well?

Comment (pabloferz) [inline test/sets.jl]: Well, it only works when the return type of the generator can be inferred.

File: base/set.jl [hunk 2]
```diff
@@ -8,6 +8,11 @@ type Set{T} <: AbstractSet{T}
 end
 Set() = Set{Any}()
 Set(itr) = Set{eltype(itr)}(itr)
+function Set(g::Generator)
+    T = _default_eltype(typeof(g))
+    T === Union{} && return Set(collect(g))
```
Comment (JeffBezanson) [inline base/set.jl]: I believe we can use `grow_to!(Set{Union{}}(), itr)` to avoid calling `collect`.

**Pull request merged**